src/lib.rs

```diff
@@ -764,44 +764,56 @@ pub enum TypeInner {
     BindingArray { base: Handle<Type>, size: ArraySize },
 }
 
-/// Constant value.
-#[derive(Debug, PartialEq)]
-#[cfg_attr(feature = "clone", derive(Clone))]
+#[derive(Debug, Clone, Copy, PartialOrd, PartialEq)]
 #[cfg_attr(feature = "serialize", derive(Serialize))]
 #[cfg_attr(feature = "deserialize", derive(Deserialize))]
 #[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
-pub struct Constant {
-    pub name: Option<String>,
-    pub specialization: Option<u32>,
-    pub inner: ConstantInner,
+pub enum Literal {
+    F64(f64),
+    U64(u64),
+    I64(i64),
+    F32(f32),
+    U32(u32),
+    I32(i32),
+    Bool(bool),
 }
 
-/// A literal scalar value, used in constants.
-#[derive(Debug, Clone, Copy, PartialOrd)]
-#[cfg_attr(feature = "serialize", derive(Serialize))]
-#[cfg_attr(feature = "deserialize", derive(Deserialize))]
-#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
-pub enum ScalarValue {
-    Sint(i64),
-    Uint(u64),
-    Float(f64),
-    Bool(bool),
+impl Literal {
+    pub const fn width(&self) -> Bytes {
+        match *self {
+            Self::F64(_) | Self::U64(_) | Self::I64(_) => 8,
+            Self::F32(_) | Self::U32(_) | Self::I32(_) => 4,
+            Self::Bool(_) => 1,
+        }
+    }
+    pub const fn scalar_kind(&self) -> ScalarKind {
+        match *self {
+            Self::F64(_) | Self::F32(_) => ScalarKind::Float,
+            Self::U64(_) | Self::U32(_) => ScalarKind::Uint,
+            Self::I64(_) | Self::I32(_) => ScalarKind::Sint,
+            Self::Bool(_) => ScalarKind::Bool,
+        }
+    }
+    fn ty_inner(&self) -> TypeInner {
+        TypeInner::Scalar {
+            kind: self.scalar_kind(),
+            width: self.width(),
+        }
+    }
 }
 
-/// Additional information, dependent on the kind of constant.
-#[derive(Clone, Debug, PartialEq)]
+/// Constant value.
+#[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "clone", derive(Clone))]
 #[cfg_attr(feature = "serialize", derive(Serialize))]
 #[cfg_attr(feature = "deserialize", derive(Deserialize))]
 #[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
-pub enum ConstantInner {
-    Scalar {
-        width: Bytes,
-        value: ScalarValue,
-    },
-    Composite {
-        ty: Handle<Type>,
-        components: Vec<Handle<Constant>>,
-    },
+pub struct Constant {
+    pub name: Option<String>,
+    pub specialization: Option<Option<u32>>,
```

Comment:
Yeah, this is a bit awkward. Perhaps:
```
enum Specialization {
    None,
    ByName,
    ByNameOrId(u32)
}
```

As a separate PR we should probably rename `specialization` to `override`.

(I'm glad the WGSL spec has continued to treat `override` as a special case of `global_constant_decl`.)